Source/santabs/SNTBundleService.m

```diff
@@ -122,170 +130,145 @@ - (void)hashBundleBinariesForEvent:(SNTStoredEvent *)event
+  // No locks are used since every file has a slot.
+  //
+  // Xcode.app has roughly 500k files, 8bytes per pointer is ~4MB for this array. This size to space
+  // ratio seems appropriate as Xcode.app is in the upper bounds of bundle size.
+  __block void **fis = calloc(dirEnum.count, sizeof(void *));
 
   // Counts used as additional progress information in SantaGUI
-  __block uint64_t binaryCount = 0;
-  __block uint64_t sentBinaryCount = 0;
-  __block uint64_t fileCount = 0;
+  __block volatile int64_t binaryCount = 0;
+  __block volatile int64_t sentBinaryCount = 0;
 
-  __block BOOL breakDir = NO;
+  // Account for 80% of the work
+  NSProgress *p;
+  if (progress) {
+    [progress becomeCurrentWithPendingUnitCount:80];
+    p = [NSProgress progressWithTotalUnitCount:dirEnum.count * 100];
+  }
 
-  // In the first stage iterate over every file in the tree checking if it is a binary. If so add
-  // it to the fis set for the second stage. Hashing the file while iterating over the filesystem
-  // causes performance issues. Do them separately.
-  while (1) {
+  // Dispatch a block for every file in dirEnum.
+  dispatch_apply(dirEnum.count, self.queue, ^(size_t i) {
     @autoreleasepool {
-      if (breakDir || progress.isCancelled) break;
-
-      // Wait for a processing thread to become available. At this stage we are only reading the
-      // mach_header. If all processing threads are blocking for more than 60 sec bail.
-      if (dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 60 * NSEC_PER_SEC))) {
-        LOGD(@"isExecutable processing threads timeout");
-        return nil;
-      }
+      if (progress.isCancelled) return;
 
-      dispatch_group_async(group,
-                           dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
-        pthread_mutex_lock(&enumeratorMutex);
-        NSString *file = [dirEnum nextObject];
-        fileCount++;
-        pthread_mutex_unlock(&enumeratorMutex);
-
-        if (!file) {
-          breakDir = YES;
-          dispatch_semaphore_signal(sema);
-          return;
-        }
-
-        if ([dirEnum fileAttributes][NSFileType] != NSFileTypeRegular) {
-          dispatch_semaphore_signal(sema);
-          return;
-        }
-
-        NSString *newFile = [event.fileBundlePath stringByAppendingPathComponent:file];
-        SNTFileInfo *fi = [[SNTFileInfo alloc] initWithPath:newFile];
-        if (!fi.isExecutable) {
-          dispatch_semaphore_signal(sema);
-          return;
-        }
-
-        pthread_mutex_lock(&eventsMutex);
-        [fis addObject:fi];
-        binaryCount++;
-        pthread_mutex_unlock(&eventsMutex);
-
-        dispatch_semaphore_signal(sema);
+      dispatch_sync(dispatch_get_main_queue(), ^{
+        p.completedUnitCount++;
       });
-      if (progress && ((fileCount % 500) == 0 || binaryCount > sentBinaryCount)) {
+
+      if (progress && ((i % 500) == 0 || binaryCount > sentBinaryCount)) {
```

Comment:
Nothing really, but it would just send an extra message to the gui. Just in case I moved this to the main thread.